File: java/src/main/java/com/workshop/architecture/fitness/managing_plans/PlanService.java
package com.workshop.architecture.fitness.managing_plans;

import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class PlanService {

    private final PlanRepository repository;

    public PlanService(PlanRepository repository) {
        this.repository = repository;
    }

    public List<PlanResponse> findAll() {
        return repository.findAll().stream()
                .sorted(Comparator.comparing(PlanEntity::getTitle))
                .map(PlanResponse::fromEntity)
                .toList();
    }

    public PlanResponse findById(UUID planId) {
        return PlanResponse.fromEntity(load(planId));
    }

    public PlanResponse create(PlanUpsertRequest request) {
        PlanEntity entity = new PlanEntity(
                UUID.randomUUID(),
                request.title(),
                request.description(),
                request.durationInMonths(),
                request.price()
        );
        return PlanResponse.fromEntity(repository.save(entity));
    }

    public PlanResponse update(UUID planId, PlanUpsertRequest request) {
        PlanEntity entity = load(planId);
        entity.updateFrom(request);
        return PlanResponse.fromEntity(repository.save(entity));
    }

    public void delete(UUID planId) {
        PlanEntity entity = load(planId);
        repository.delete(entity);
    }

    private PlanEntity load(UUID planId) {
        return repository.findById(planId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Plan %s was not found".formatted(planId)
                ));
    }
}

File: python/src/workshop_api/fitness/business/application/plan_service.py
from uuid import UUID

from workshop_api.fitness.business.application.errors import NotFoundError
from workshop_api.fitness.infrastructure.plan_entity import PlanOrmModel
from workshop_api.fitness.infrastructure.plan_repository import PlanRepository
from workshop_api.fitness.presentation.plan_schemas import PlanResponse, PlanUpsertRequest


class PlanService:
    def __init__(self, repository: PlanRepository) -> None:
        self.repository = repository

    def list_plans(self) -> list[PlanResponse]:
        plans = sorted(self.repository.find_all(), key=lambda plan: plan.title)
        return [self._to_response(plan) for plan in plans]

    def get_plan(self, plan_id: UUID) -> PlanResponse:
        return self._to_response(self._load_plan(plan_id))

    def create_plan(self, request: PlanUpsertRequest) -> PlanResponse:
        plan = PlanOrmModel(
            title=request.title,
            description=request.description,
            duration_in_months=request.duration_in_months,
            price=request.price,
        )
        return self._to_response(self.repository.save(plan))

    def update_plan(self, plan_id: UUID, request: PlanUpsertRequest) -> PlanResponse:
        plan = self._load_plan(plan_id)
        plan.title = request.title
        plan.description = request.description
        plan.duration_in_months = request.duration_in_months
        plan.price = request.price
        return self._to_response(self.repository.save(plan))

    def delete_plan(self, plan_id: UUID) -> None:
        plan = self._load_plan(plan_id)
        self.repository.delete(plan)

    def _load_plan(self, plan_id: UUID) -> PlanOrmModel:
        plan = self.repository.find_by_id(plan_id)
        if plan is None:
            raise NotFoundError(f"Plan {plan_id} was not found")
        return plan

    @staticmethod
    def _to_response(plan: PlanOrmModel) -> PlanResponse:
        return PlanResponse(
            id=plan.id,
            title=plan.title,
            description=plan.description,
            durationInMonths=plan.duration_in_months,
            price=plan.price,
        )
